src/libraries/Common/src/System/Threading/AsyncOverSyncWithIoCancellation.cs

```diff
@@ -150,64 +193,33 @@ public void Dispose()
-                if (WorkItem.CallbackCompleted is Task t)
-                {
-                    await t.ConfigureAwait(false);
-                }
+                WorkItem._callbackCompleted?.GetAwaiter().GetResult();
             }
         }
 
         /// <summary>Registers for cancellation with the specified token.</summary>
         /// <remarks>Upon cancellation being requested, the implementation will attempt to CancelSynchronousIo for the thread calling RegisterCancellation.</remarks>
-        private static SyncAsyncWorkItemRegistration RegisterCancellation(CancellationToken cancellationToken) =>
-            cancellationToken.CanBeCanceled ? RegisterCancellation(new AsyncOverSyncWithIoCancellation(), cancellationToken) :
-            default; // If the token can't be canceled, there's nothing to register.
-
-        /// <summary>Registers for cancellation with the specified token.</summary>
-        /// <remarks>Upon cancellation being requested, the implementation will attempt to CancelSynchronousIo for the thread calling RegisterCancellation.</remarks>
-        private static SyncAsyncWorkItemRegistration RegisterCancellation(AsyncOverSyncWithIoCancellation instance, CancellationToken cancellationToken)
+        private static SyncAsyncWorkItemRegistration RegisterCancellation(CancellationToken cancellationToken)
         {
-            // Get a handle for the current thread. This is stored and used to cancel the I/O on this thread
-            // in response to the cancellation token having cancellation requested.  If the handle is invalid,
-            // which could happen if OpenThread fails, skip attempts at cancellation. The handle needs to be
-            // opened with THREAD_TERMINATE in order to be able to call CancelSynchronousIo.
-            instance.ThreadHandle = t_currentThreadHandle;
-            if (instance.ThreadHandle is null)
+            if (!cancellationToken.CanBeCanceled)
             {
-                instance.ThreadHandle = Interop.Kernel32.OpenThread(Interop.Kernel32.THREAD_TERMINATE, bInheritHandle: false, Interop.Kernel32.GetCurrentThreadId());
-                if (instance.ThreadHandle.IsInvalid)
-                {
-                    int lastError = Marshal.GetLastPInvokeError();
-                    Debug.Fail($"{nameof(Interop.Kernel32.OpenThread)} unexpectedly failed with 0x{lastError:X8}: {Marshal.GetPInvokeErrorMessage(lastError)}");
-                    return default;
-                }
+                return default;
+            }
 
-                t_currentThreadHandle = instance.ThreadHandle;
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Get the instance for this thread. If the instance doesn't have a thread handle, that
+            // means we were unable to obtain one previously, and we shouldn't try to cancel I/O.
+            AsyncOverSyncWithIoCancellation instance = t_instance ??= new AsyncOverSyncWithIoCancellation();
```

Comment:
really clever resources reuse (ThreadStatic so per thread, sync so no need for synchronization as one thread can do only one sync thing at a time) 👍